<a href="https://colab.research.google.com/github/biniwollo/Homework2/blob/main/DASHomeworkTwo_Ali.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Football Coach Salary Analysis.This code performs an analysis of football coach salaries in college football. 
It explores the relationship between coach salaries and various factors such as graduation rates and team performance 
statistics. The code retrieves data on coaches, stadiums, and statistics, merges them into a single data frame, and 
conducts data preprocessing steps. It then calculates summary statistics, visualizes the data distribution, and examines 
correlations between variables. The code fits a linear regression model to predict coach salaries based on graduation rates 
and team performance. It evaluates the model's performance and uses it to answer specific questions, such as recommended salaries 
and the impact of moving to a different conference. The code provides insights into the factors influencing coach salaries and offers 
meaningful answers to the questions posed in the analysis. It utilizes libraries such as pandas, numpy, seaborn, matplotlib, statsmodels,
and scikit-learn.

In [ ]:
#DSAHomeworkTwo

# Author: [Binyam Ali]
# Date: [06/10/2023]

import numpy as np
import pandas as pd

# Obtain Data

coaches = pd.read_csv('https://raw.githubusercontent.com/biniwollo/Homework2/main/coaches_bp_dsa.csv')
stadium = pd.read_csv('https://raw.githubusercontent.com/biniwollo/Homework2/main/stadium_bp_dsa.csv')
stats = pd.read_csv('https://raw.githubusercontent.com/biniwollo/Homework2/main/stats_bp_dsa.csv')

print(coaches)
#print(stadium)
#print(stats)



**Data cleaning and preparatioin:**
Handling missing values:
check for missing values in the data frame




In [ ]:
# Scrub Data

# Handle missing values in coaches data frame
coaches.dropna(inplace=True)  # Drop rows with missing values

# Handle missing values in stadium data frame
stadium.dropna(inplace=True)  # Drop rows with missing values

# Handle missing values in stats data frame
stats.fillna(0, inplace=True)  # Fill missing values with 0

# Merge data frames
merged_data = pd.merge(pd.merge(coaches, stadium, on='School'), stats, on='School')

# Display the merged data frame
print("\nMerged data frame ------------------")
print(merged_data)

**Data cleaning and preparatioin:**
Convert data types:
Ensure that the columns have the correct data type for analysis

In [ ]:
# Scrub Data

# Remove leading and trailing spaces from column names
merged_data.columns = merged_data.columns.str.strip()

# Remove commas and dollar signs from the "TotalPay" column
merged_data['TotalPay'] = merged_data['TotalPay'].str.replace(',', '').str.replace('$', '')

# Convert the "TotalPay" column to integer
merged_data['TotalPay'] = merged_data['TotalPay'].astype(int)

# Display the updated data frame
print("\nUpdated Merged data frame ------------------")
print(merged_data)



Summary statistics code uses the describe() function of Pandas, which provides summary statistics for numeric columns in a data frame. 
It includes count, mean, standard deviation,minimum value, 
25th percentile,median (50th percentile), 
75th percentile, and maximum value.

when we run this code after merging and updating the data frame, it will calculate and display the summary statistics for the numeric columns in the merged_data data frame.


In [ ]:
# Explore Data
# Summary statistics


summary_stats = merged_data.describe()

# Display the summary statistics
print("\nSummary Statistics ------------------")
print(summary_stats)


In [ ]:
#Explore Data - Compute value counts for a specific column

# Compute value counts for a Conf_x column
value_counts = merged_data['Conf_x'].value_counts()

# Display the value counts
print(value_counts)

In [ ]:
#Explore Data - Compute value counts for a specific column

# Compute value counts for a TotalPay column
value_counts = merged_data['TotalPay'].value_counts()

# Display the value counts
print(value_counts)

In [ ]:
#Explore Data  - plot distribution of TotalPay

import seaborn as sns
import matplotlib.pyplot as plt

# Distribution plot for TotalPay
sns.histplot(merged_data['TotalPay'], kde=True)
plt.title('Distribution of TotalPay')
plt.xlabel('TotalPay')
plt.ylabel('Frequency')
plt.show()

In [ ]:
#Explore Data  - plot distribution of Graduation Rate(GSR)

# Distribution plot for Graduation Rate (GSR)
sns.histplot(merged_data['Graduation Rate (GSR)'], kde=True)
plt.title('Distribution of Graduation Rate (GSR)')
plt.xlabel('Graduation Rate (GSR)')
plt.ylabel('Frequency')
plt.show()

In [ ]:
#Explore Data  - scatter plot TotalPay VS. Graduation Rate(GSR)

# Scatter plot: TotalPay vs. Graduation Rate (GSR)
sns.scatterplot(data=merged_data, x='TotalPay', y='Graduation Rate (GSR)')
plt.title('TotalPay vs. Graduation Rate (GSR)')
plt.xlabel('TotalPay')
plt.ylabel('Graduation Rate (GSR)')
plt.show()


In [ ]:
#Explore Data  - scatter plot PointsPerGame VS. Score

# Scatter plot: PointsPerGame vs. Score
sns.scatterplot(data=merged_data, x='PointsPerGame', y='Score')
plt.title('PointsPerGame vs. Score')
plt.xlabel('PointsPerGame')
plt.ylabel('Score')
plt.show()

In [ ]:
#Explore - Correleation between TotalPay, Graduation Rate (GSR), PointsPerGame and Score

# Correlation matrix
correlation_matrix = merged_data[['TotalPay', 'Graduation Rate (GSR)', 'PointsPerGame', 'Score']].corr()
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm')
plt.title('Correlation Matrix')
plt.show()

Fit a regression model with the salary as the response variable:

We use of the statsmodels library to perform a linear regression analysis. 

In [ ]:
import statsmodels.api as sm

# Define the dependent variable
y = merged_data['TotalPay']

# Define the independent variables
X = merged_data[['Graduation Rate (GSR)', 'PointsPerGame']]

# Add a constant term to the independent variables
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(y, X).fit()

# Print the model summary
print(model.summary())



The following code builds a linear regression model using the scikit-learn library. It splits the data into training and test sets, trains the model, makes predictions, and evaluates its performance using mean squared error and R-squared metrics.

In [94]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Define the dependent variable
y = merged_data['TotalPay']

# Define the independent variables
X = merged_data[['Graduation Rate (GSR)', 'PointsPerGame']]

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a linear regression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print the evaluation metrics
print("Mean Squared Error:", mse)
print("R-squared:", r2)

Mean Squared Error: 2516410876269.1553
R-squared: -0.18483227315930328


**Question 1:**
What is the recommended salary for the OU football coach? 

**Note:**

According to a Google search, the graduation rate for OU in the 2020-21 academic year was 67.5%. Additionally, the average points per game in 2021 were 36. By utilizing these two figures in a linear regression model, a recommended salary for the OU football coach was predicted

In [95]:
# Question 1
# Assume GSR and PointsPerGame values
gsr = 67.5
ppg = 36

# Create a new data point with the assumed values
new_data = pd.DataFrame({'Graduation Rate (GSR)': [gsr], 'PointsPerGame': [ppg]})

# Predict the salary using the regression model
predicted_salary = model.predict(new_data)

# Print the predicted salary
print("Recommended Salary: $", predicted_salary[0])

Recommended Salary: $ 1568727.5348648978


**Answer:**
The recommended salary for the OU football coach *$ 1,568,727.53.*

**Question 2:**
What would the appropriate salary be if OU moved to the SEC?

**Note:**
To determine the appropriate salary for the OU football coach if they moved to the SEC, first, we will determine the average graduation rate (GSR) and Points Per Game in the SEC. Then, we can use these values to predict the salary using the regression model created above.

In [ ]:
# To determine the average graduation rate (GSR) and Points Per Game in the SEC

# Compute the average for a specific column

average_values = merged_data.groupby('Conf_x')['PointsPerGame'].mean()
# Display the average values
print("The average Points Per Game by conferecne:", average_values)

#The verage Points Per Game in the SEC = 72.86

average_values = merged_data.groupby('Conf_x')['Graduation Rate (GSR)'].mean()
# Display the average values
print("The average Graduation Rate by conferecne:", average_values)

#The verage graduation rate (GSR)SEC = 75.77

In [97]:
#To determine the appropriate salary for the OU football coach if they moved to the SEC

# average GSR and PointsPerGame values for the SEC
avg_gsr_sec = 75.77
avg_ppg_sec = 72.86

# Create a new data point with the assumed SEC values
sec_data = pd.DataFrame({'Graduation Rate (GSR)': [avg_gsr_sec], 'PointsPerGame': [avg_ppg_sec]})

# Predict the salary using the regression model
predicted_salary_sec = model.predict(sec_data)

# Print the predicted salary for the SEC scenario
print("Appropriate Salary in SEC: $", predicted_salary_sec[0])



Appropriate Salary in SEC: $ 2783310.0078133917


**Answer:**


Answer: 
The recommended salary for the OU football coach if OU moved to the SEC is $ 2,783,310.00.



**Question 3:** What schools did we drop from our data and why?

To show the dropped schools after handling missing values in the coaches data frame, we can compare the original coaches data frame with the updated merged data frame.

In [104]:
#comparing the orginal coaches data to the updated merged data frame

# ... Previous code ...

# Create a list of dropped schools
dropped_schools = list(set(coaches['School']) - set(merged_data['School']))

# Display the dropped schools
print("\nDropped Schools:")
for school in dropped_schools:
    print(school)



Dropped Schools:
Temple
Brigham Young
Southern Methodist
Baylor
Tennessee


**Answer:**
We have dropped the following schools


1. Temple
2. Brigham Young
3. Southern Methodist
4. Baylor
5. Tennessee


The schools might have been dropped because they had missing values in either the coaches data frame, stadium data frame, or both. The dropna() function is used to remove rows that contain missing values from a data frame.

If a school had missing values in either the coaches data frame or stadium data frame, it would be dropped during the merge operation because the merge is performed based on the 'School' column. Only the schools that have complete data in both data frames will be included in the merged_data data frame.

This approach ensures that the resulting merged_data data frame only contains data for schools that have complete information in terms of coaches, stadiums, and stats.






**Question 4:** What is the single biggest impact on salary size?

**Note:**
To determine the single biggest impact on salary size, you can look at the coefficient values from the linear regression model. The coefficient represents the change in the dependent variable (salary) for a unit change in the corresponding independent variable while holding other variables constant.

In [106]:
# Extract the coefficients from the linear regression model
coefficients = model.coef_

# Get the feature names from the X dataframe
feature_names = X.columns

# Create a dictionary with variable names and their coefficients
coefficients_dict = dict(zip(feature_names, coefficients))

# Sort the coefficients in descending order based on their absolute values
sorted_coefficients = sorted(coefficients_dict.items(), key=lambda x: abs(x[1]), reverse=True)

# Print the variable with the largest impact on salary size
print("Variable with the largest impact on salary size:")
print(sorted_coefficients[0])

Variable with the largest impact on salary size:
('PointsPerGame', 28196.357091131875)


**Answer: **

The variable with the single biggest impact on salary sizeis '**PointsPerGame**'. Its coefficient value is *28196.357091131875*, indicating that for every unit increase in 'PointsPerGame', the salary size is expected to increase by approximately 28196.36 units. 

Therefore, '**PointsPerGame'** has the largest positive impact on the salary size in the linear regression model.
